src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,758 @@
+        {
+            return HubProtocolExtensions.GetMessageBytes(this, message);
+        }
+
+        private HubMessage ParseMessage(ReadOnlySequence<byte> input, IInvocationBinder binder)
+        {
+            try
+            {
+                // We parse using the Utf8JsonReader directly but this has a problem. Some of our properties are dependent on other properties
+                // and since reading the json might be unordered, we need to store the parsed content as JsonDocument to re-parse when true types are known.
+                // if we're lucky and the state we need to directly parse is available, then we'll use it.
+
+                int? type = null;
+                string invocationId = null;
+                string target = null;
+                string error = null;
+                var hasItem = false;
+                object item = null;
+                var hasResult = false;
+                object result = null;
+                var hasArguments = false;
+                object[] arguments = null;
+                string[] streamIds = null;
+                JsonDocument argumentsToken = null;
+                JsonDocument itemsToken = null;
+                JsonDocument resultToken = null;
+                ExceptionDispatchInfo argumentBindingException = null;
+                Dictionary<string, string> headers = null;
+                var completed = false;
+
+                var reader = new Utf8JsonReader(input, isFinalBlock: true, state: default);
+
+                reader.CheckRead();
+
+                // We're always parsing a JSON object
+                reader.EnsureObjectStart();
+
+                do
+                {
+                    switch (reader.TokenType)
+                    {
+                        case JsonTokenType.PropertyName:
+                            if (reader.TextEquals(TypePropertyNameBytes))
+                            {
+                                type = reader.ReadAsInt32(TypePropertyName);
+
+                                if (type == null)
+                                {
+                                    throw new InvalidDataException($"Missing required property '{TypePropertyName}'.");
+                                }
+                            }
+                            else if (reader.TextEquals(InvocationIdPropertyNameBytes))
+                            {
+                                invocationId = reader.ReadAsString(InvocationIdPropertyName);
+                            }
+                            else if (reader.TextEquals(StreamIdsPropertyNameBytes))
+                            {
+                                reader.CheckRead();
+
+                                if (reader.TokenType != JsonTokenType.StartArray)
+                                {
+                                    throw new InvalidDataException($"Expected '{StreamIdsPropertyName}' to be of type {reader.GetTokenString()}.");
```

Comment:
Should `reader.GetTokenString()` be `SystemTextJsonExtensions.GetTokenString(JsonTokenType.StartArray)`?